### PR TITLE
[DO NOT MERGE] -ETOOMANYCONFIGS

### PR DESCRIPTION
**Pull request by rchiossi**
This is the first 100% test coverage version of the config unification patch.

The first big change with this patch is the move of config from builder to its own package. While inside builder it could not be used inside swupd since that would generate a cyclic dependency. Moving it outside allows for reuse in both parts of the code.

Another change is the rename of a bunch of stuff called config. With config now in its own package, it had conflicts with the config object defined inside swupd and the config global variable used to hold the config file name in the cmd package. Both are now renamed properly.

Note that most of this patch is focused on removing the extra config files (namely mixversion, upstreamversion, upstreamurl and server.ini) and not in reworking the communication API between builder and swupd. That part still need some work to remove redundant parameter passing. Since that rework is also a big patch, I decided to split the PR and send that on top of this one.

Also, note that while test coverage is 100%, that does not include use with --new-config, so there might be crashes that sliped through my review. Adding --new-config to tests is something that I'm evaluating atm and wasn't included mostly because I don't wanna duplicate all tests just to remove then later when new config becomes default.

### CODE REVIEW

**Comment by matthewrsj** [comment]
> Adding --new-config to tests is something that I'm evaluating atm and wasn't included mostly because I don't wanna duplicate all tests just to remove then later when new config becomes default.

We should definitely add --new-config to the bat tests. It is completely fine to remove the flag later. We need to be testing the --new-config option since we want to be going to that implementation as our default eventually.

File: builder/builder.go [hunk 1]
```diff
@@ -1274,13 +1272,23 @@ func (b *Builder) ValidateBundles(bundles []string, lvl ValidationLevel) error {
...
 func (b *Builder) UpdateMixVer() error {
-	// Deprecate '.mixversion' --> 'mixversion'
-	if _, err := os.Stat(filepath.Join(b.Config.Builder.VersionPath, ".mixversion")); err == nil {
-		b.MixVerFile = ".mixversion"
-		fmt.Println("Warning: '.mixversion' has been deprecated. Please rename file to 'mixversion'")
+	mixVer, _ := strconv.Atoi(b.Config.Builder.MixVer)
+	b.Config.Builder.MixVer = strconv.Itoa(mixVer + 10)
+
+	if !config.UseNewConfig {
+		// Deprecate '.mixversion' --> 'mixversion'
```
Comment (matthewrsj) [inline builder/builder.go]: I've noticed that these deprecation codepaths only exist under the `!config.UseNewConfig` guard, when users go to `--new-config` we can expect them not to need the deprecation warning.

But what about when we get rid of `--new-config`? Is that the end of the deprecation period and we will stop doing those checks? Or will we have to re-introduce the deprecation codepaths to the regular codepath?

File: swupd/data.go [hunk 2]
```diff
@@ -31,79 +32,32 @@ type dbgConfig struct {
 	src    string
 }
 
-type config struct {
-	stateDir  string
+type swupdData struct {
```
Comment (matthewrsj) [inline swupd/data.go]: This is a config, not data. Can you rename this appropriately?

File: swupd/data_test.go [hunk 3]
```diff
@@ -5,29 +5,6 @@ import (
 	"testing"
 )
 
-func TestReadServerINI(t *testing.T) {
```
Comment (matthewrsj) [inline swupd/data_test.go]: You remove this test here because the `swupd` package doesn't read the config anymore. Can you add a similar one to where the config is actually being read?

**Comment by matthewrsj** [comment]
Also please run `make lint` against the codebase.

File: swupd/data.go (hunk 2 above)
Comment (rchiossi) [reply swupd/data.go]: Yeah, i know. I just dont want to call if config. That is exactly what I`ve been trying to remove all along do avoid confusion. Actually, I want to get rid of this file and keep config/config.go only, but that requires some more refactoring.

File: swupd/data_test.go (hunk 3 above)
Comment (rchiossi) [reply swupd/data_test.go]: What do you mean? To create a new test file for config module? I removed this test because what it tests does not exists anymore.

File: builder/builder.go (hunk 1 above)
Comment (rchiossi) [reply builder/builder.go]: When we remove --new-config, is this file exists there it will just be ignored. If by then the user hasn' t upgraded to the new format, that will be the least of his problems. All the rest will also not work.

**Comment by rchiossi** [comment]
This PR conflicts with my own PR, but the other should go in first. Also there are some other changes to this patch that will add further modification. I'll wait for the other to get in before updating this PR.

File: builder/builder.go (hunk 1 above)
Comment (kevincwells) [reply builder/builder.go]: I think I'm missing something. The deprecation warnings @matthewrsj is talking about are for people using the old names for the version files (e.g., `.clearurl` -> `upstreamurl`). Are we having a separate deprecation period for people using the version files at all?

I haven't read this whole PR thoroughly yet, but am I right that (if you're using `--new-config`) these files are just ignored altogether immediately? Should we have a period of time where, if these files exist, a warning is output, but the value in them is still used?

File: builder/builder.go [hunk 4]
```diff
@@ -200,54 +184,65 @@ func (b *Builder) InitMix(upstreamVer string, mixVer string, allLocal bool, allU
...
 
-	// Deprecate '.clearversion' --> 'upstreamversion'
-	if _, err := os.Stat(filepath.Join(b.Config.Builder.VersionPath, ".clearversion")); err == nil {
-		b.UpstreamVerFile = ".clearversion"
-		fmt.Println("Warning: '.clearversion' has been deprecated. Please rename file to 'upstreamversion'")
-	}
-	if err := ioutil.WriteFile(filepath.Join(b.Config.Builder.VersionPath, b.UpstreamVerFile), []byte(upstreamVer), 0644); err != nil {
-		return err
+		if config.UseNewConfig {
+			if err := b.Config.SaveConfig(b.BuildConf); err != nil {
```
Comment (kevincwells) [inline builder/builder.go]: Not understanding yet why this call is inside the `if b.Config.Builder.UpstreamVer == "latest"` block. Don't you need to write back out the config either way after overwriting all the `b.Config.Builder.*` objects a few lines above?

**Comment by rchiossi** [comment]
Closing as this has been completely reworked. PR #321 is the first part of the replacement PR.
